import "distributions/io/schema.proto";

package protobuf.loom;

//----------------------------------------------------------------------------
// Products

message ProductModel {
  message Shared {
    required distributions.Clustering clustering = 1;
    repeated distributions.BetaBernoulli.Shared bb = 2;
    repeated distributions.DirichletDiscrete.Shared dd = 3; // ordered by increasing dim
    repeated distributions.DirichletProcessDiscrete.Shared dpd = 4;
    repeated distributions.GammaPoisson.Shared gp = 5;
    repeated distributions.NormalInverseChiSq.Shared nich = 6;
  }

  message HyperPrior {
    repeated distributions.Clustering clustering = 1;
    required distributions.BetaBernoulli.GridPrior bb = 2;
    required distributions.DirichletDiscrete.GridPrior dd = 3;
    required distributions.DirichletProcessDiscrete.GridPrior dpd = 4;
    required distributions.GammaPoisson.GridPrior gp = 5;
    required distributions.NormalInverseChiSq.GridPrior nich = 6;
  }

  message SparseValue {
    repeated bool observed = 1;
    repeated bool booleans = 2;  // including bb
    repeated uint32 counts = 3;  // including dd, dpd, gp
    repeated float reals = 4;  // including nich
  }

  message Group {
    required uint64 count = 1;
    repeated distributions.BetaBernoulli.Group bb = 2;
    repeated distributions.DirichletDiscrete.Group dd = 3;
    repeated distributions.DirichletProcessDiscrete.Group dpd = 4;
    repeated distributions.GammaPoisson.Group gp = 5;
    repeated distributions.NormalInverseChiSq.Group nich = 6;
  }
}

//----------------------------------------------------------------------------
// CrossCat

message CrossCat {
  message Kind {
    required ProductModel.Shared product_model = 1;
    repeated uint32 featureids = 2;
  }

  message HyperPrior {
    required ProductModel.HyperPrior inner_prior = 1;
    repeated distributions.Clustering outer_prior = 2;
  }

  repeated Kind kinds = 1;
  required distributions.Clustering feature_clustering = 2;
  repeated uint32 featureid_to_kindid = 3;
  optional HyperPrior hyper_prior = 4;
}

message SparseRow {
  required uint64 id = 1;
  required ProductModel.SparseValue data = 2;
}

message Assignment {
  required uint64 rowid = 1;
  repeated uint32 groupids = 2;
}

message PosteriorEnum {
  message Group {
    repeated uint32 rowids = 1;
  }
  message Kind {
    repeated uint32 featureids = 1;
    repeated Group groups = 2;
  }
  message Sample {
    repeated Kind kinds = 1;
    optional float score = 2;
  }
}

//----------------------------------------------------------------------------
// Infer Log

message InferLog
{
  message Rusage
  {
    required uint64 max_resident_size_kb = 1;
    required double user_time_sec = 2;
    required double sys_time_sec = 3;
  }

  message Args
  {
    message Timer
    {
      required string name = 1;
      required float elapsed = 2;
    }
    message Summary
    {
      message KindHypers
      {
        repeated float alphas = 1;
        repeated float ds = 2;
      }
      message ModelHypers
      {
        required float alpha = 1;
        required float d = 2;
      }

      required ModelHypers model_hypers = 1;
      required KindHypers kind_hypers = 2;
      repeated uint32 feature_counts = 3;
      repeated uint32 category_counts = 4;
    }
    message Scores
    {
      optional float score = 1;
      optional float kl_divergence = 2;
      optional uint64 total_object_count = 3;
      optional uint64 assigned_object_count = 4;
      repeated float features = 5;
    }
    message KernelStatus
    {
      message Algo8
      {
        required uint64 total_count = 1;
        required uint64 change_count = 2;
      }

      optional Algo8 algo8 = 1;
    }

    optional uint32 iter = 1;
    repeated Timer timers = 2;
    optional Summary summary = 3;
    optional Scores scores = 4;
    optional KernelStatus kernel_status = 5;
  }

  required uint64 timestamp_usec = 1;
  required Rusage rusage = 2;
  required Args args = 3;
}

//----------------------------------------------------------------------------
// PreQL

message PreQL {
  message Predict {
    message Query {
      required string id = 1;
      required ProductModel.SparseValue data = 2;
      repeated bool to_predict = 3;
      required uint32 sample_count = 4;
    }
    message Result {
      required string id = 1;
      optional string error = 2;
      repeated ProductModel.SparseValue samples = 3;
    }
  }
}
